bigframes/core/tree_properties.py

```diff
@@ -32,14 +32,45 @@ def local_only(node: nodes.BigFrameNode) -> bool:
     return all(isinstance(node, nodes.ReadLocalNode) for node in node.roots)
 
 
-def peekable(node: nodes.BigFrameNode) -> bool:
+def can_fast_peek(node: nodes.BigFrameNode) -> bool:
     if local_only(node):
         return True
-    children_peekable = all(peekable(child) for child in node.child_nodes)
+    children_peekable = all(can_fast_peek(child) for child in node.child_nodes)
     self_peekable = not node.non_local
     return children_peekable and self_peekable
 
 
+def can_fast_head(node: nodes.BigFrameNode) -> bool:
+    """Can get head fast if can push head operator down to leafs and operators preserve rows."""
+    if isinstance(node, nodes.LeafNode):
+        return node.supports_fast_head
+    # TODO: In theory we can push head down through concat, but requires some dedicated logic
```

Comment:
Are you thinking something like: limit applied to each + an overall limit clause? Does the BigQuery engine not push down `limit` through a union if we apply it at the top level?

Reply:
I guess I didn't think too hard about this. I am not aware of engine-level ordered head optimization that would prevent dispatching all the underlying data